Verify Justified Block Hash

JerryFireman opened the pull request:
TODO(you): choose "part of" or "resolves" and the associated github issue #.

Resolves #721 

---

# Description

TODO(you): Write why you are making the changes in this pull request.
Verify that the attestation's slot number is the same as the slot number of the block retrieved by the justified slot hash.

TODO(you): Write a summary of the changes you are making.
Added code to return error if the the attestation's slot number does not match the slot number of the block retrieved by the justified slot hash.


TODO(you): Link anything that would be helpful or relevant to the reviewers.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/prysmaticlabs/prysm?pullRequest=747) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/prysmaticlabs/prysm?pullRequest=747) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/prysmaticlabs/prysm?pullRequest=747) it.</sub>

terenc3t commented on .idea/vcs.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Can we not check this in? Thanks

terenc3t commented on beacon-chain/types/block.go:
```diff
@@ -271,6 +272,19 @@ func (b *Block) isAttestationValid(attestationIndex int, db beaconDB, aState *Ac
 		return false
 	}
 
+	// Retrieve block from db using justified_block_hash and verify its slot number is the same as attestation justified slot number
```
please add a period at the end of the comment

terenc3t commented on beacon-chain/types/block.go:
```diff
@@ -271,6 +272,19 @@ func (b *Block) isAttestationValid(attestationIndex int, db beaconDB, aState *Ac
+	// Retrieve block from db using justified_block_hash and verify its slot number is the same as attestation justified slot number
+	rBlock, err := db.GetBlock(hash)
+	if err != nil {
+		log.Errorf("failed to get block: %v", err)
```
shouldn't you also return false here?

terenc3t replied:
Capitalize `failed`

terenc3t commented on beacon-chain/types/block.go:
```diff
@@ -271,6 +272,19 @@ func (b *Block) isAttestationValid(attestationIndex int, db beaconDB, aState *Ac
+		log.Errorf("failed to get block: %v", err)
+	}
+	if attestation.JustifiedSlot != rBlock.SlotNumber() {
+		log.Errorf("the attestation's slot number is not the same as the slot number of the block retrieved by the justified slot hash. Attestation slot number: %v",
```
Capitalize `the`

JerryFireman commented:
Hi Terence

I believe I have addressed all of your comments and updated the PR.

The tests passed but I noticed some errors in the types/block.go file in areas that I have not touched. I am assuming that someone else is making changes to the file. I hope it was not something that I did.

Please let me know how the updated PR looks and if there are any more issues for me to address.

Jerry

terenc3t commented:
@JerryFireman I don't think those comments are addressed, can you double check?

terenc3t commented on .idea/misc.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I don't think we should be checking `.idea/` in

rauljordan replied:
+1

nisdas replied:
+1 , please remove the `.idea/` folder

prestonvanloon replied:
Feel free to add `.idea` to `.gitignore` to preview this in the future

rauljordan commented on beacon-chain/types/block.go:
```diff
@@ -271,6 +272,20 @@ func (b *Block) isAttestationValid(attestationIndex int, db beaconDB, aState *Ac
 		return false
 	}
 
+	// Retrieve block from db using justified_block_hash and verify its slot number is the same as attestation justified slot number.
```
Line too long, please split for more readability

nisdas replied:
+1

rauljordan commented on beacon-chain/types/block.go:
```diff
@@ -271,6 +272,20 @@ func (b *Block) isAttestationValid(attestationIndex int, db beaconDB, aState *Ac
+		return false
+	}
+	if attestation.JustifiedSlot != rBlock.SlotNumber() {
+		log.Errorf("The attestation's slot number is not the same as the slot number of the block retrieved by the justified slot hash. Attestation slot number: %v",
```
see above comment ^

nisdas commented on beacon-chain/types/block.go:
```diff
@@ -271,6 +272,20 @@ func (b *Block) isAttestationValid(attestationIndex int, db beaconDB, aState *Ac
+	if attestation.JustifiedSlot != rBlock.SlotNumber() {
+		log.Errorf("The attestation's slot number is not the same as the slot number of the block retrieved by the justified slot hash. Attestation slot number: %v",
+			attestation.JustifiedSlot)
+		log.Errorf("Retrieved block slot number: %v",
```
This log is short enough to be put in one line, it doesnt have to be split into two.

nisdas replied:
Also the formatting for slot number should be `%d` not `%v`, the same applies in the log above

prestonvanloon commented:
I've removed the `.idea` directory for you @JerryFireman :)

prestonvanloon commented on beacon-chain/types/block.go:
```diff
@@ -271,6 +272,21 @@ func (b *Block) isAttestationValid(attestationIndex int, db beaconDB, aState *Ac
+		log.Errorf("The attestation's slot number is not the same as the slot number of the block retrieved by the justified slot hash. " +
+			"Attestation slot number: %d",
+			attestation.JustifiedSlot)
+		log.Errorf("Retrieved block slot number: %d",rBlock.SlotNumber())
```
Please make this part of the error message on the previous line. This is one issue and only one error should be logged

JerryFireman replied:
I think I resigned the CLA. If it did not work, please let me know and I will try again.

prestonvanloon commented:
I think you need to sign the CLA again. The bot dropped all of the data a little while ago. Sorry about that!

JerryFireman commented:
By the way, do you see any value in adding more detailed instructions for working with git/github to your contributor's guidelines? I am far from an expert but I have been reading up and would be happy to take a stab at it if you would like.

terenc3t commented:
> By the way, do you see any value in adding more detailed instructions for working with git/github to your contributor's guidelines? I am far from an expert but I have been reading up and would be happy to take a stab at it if you would like.

We would love that! Feel free to open up a PR. Thank you

JerryFireman commented:
@terenc3t OK I will get working on it. By the way, should I open the PR now or wait until I have something ready for review?

terenc3t commented:
> @terenc3t OK I will get working on it. By the way, should I open the PR now or wait until I have something ready for review?

When you have something for review. Thanks!

JerryFireman commented:
OK will do. Thanks for clarifying.

rauljordan commented:
Hi @JerryFireman do you have an update on this PR?

JerryFireman commented:
Hi @rauljordan, I was thinking that where we left off Preston had one change and said he would make it hmself. But I am guessing he is busy so I will go ahead and do it. Will try to do it today, tomorrow at the latest.

JerryFireman commented:
Hi @rauljordan @prestonvanloon I addressed Preston's latest comment by eliminating the redundant error and combining its message with the previous line. I believe I have now addressed all of the comments. Please let me know if I missed any or if you see any other areas that can be improved.

JerryFireman commented:
@terenc3t, per our discussion I created a new pull request to update CONTRIBUTING.md with more detailed Git and Github workflow.

Also, @prestonvanloon did I correctly address your comment from 17 days ago on this PR with my change from 12 days ago? Please let me know if there is anything else for me to do.

rauljordan commented:
Hey @JerryFireman, unfortunately a lot has changed in the spec related to this PR and we have been working on refactoring our code to align with latest changes. This functionality is out of sync now with latest spec requirements and would be best to close the PR for now. Thanks for your help on this and we'll do a better job at keeping you in the loop on which functions are least likely to change moving forward.
